### PR TITLE
Give a meaningful error when doctest has invalid prompt

### PR DESCRIPTION
**Pull request by eksperimental**
Launching iex with:

  $ iex --sname foo

The following code will break doctests

```
  iex(foo@HOST)1> :foo
  :foo
```

and will give a non descriptive error.
That's because "skip_iex_number" never found the ")>" piece, so it
consumes the line all the way to the end until it's empty.

### CODE REVIEW

**Comment by eksperimental** [comment]
PR originally opened here: https://github.com/elixir-lang/elixir/pull/5385

**Comment by whatyouhide** [comment]
Could we instead raise a descriptive message, but still not support the host? It's easier to not support something and add support later (when a legit use case comes up) than to support something and then drop support for it.

**Comment by josevalim** [comment]
Yes, we should change `skip_iex_number` to raise a nice error message if we get to the end of the string.

**Comment by eksperimental** [comment]
working on it.

one question: is this a valid test?

```elixir
  @doc """
      iex(1)> 1 +
      ...> 2
      3
  """
```

or should the second line be ?
`      ...(1)> 2`

**Comment by josevalim** [comment]
@eksperimental i don't think we care about the precise notation. I am also do not think we should. You could even have the numbers out of order. I would say that's more of a documentation bug than a doctest bug.

**Comment by eksperimental** [comment]
so before I start working on this, i want to know if there is a will to add the following checks and whether they are valid or not.

1) first line numbered, continued line not numbered
```elixir
      iex(1)> 1 +
      ...> 2
```
2) first line not numbered, continued line numbered
```elixir
      iex> 1 +
      ...(1)> 2
```
3) number in continued line does not match 
```elixir
      iex(1)> 1 +
      ...(2)> 2
```

4) first line starts with `...>`
```elixir
      ...> 2
```

**Comment by josevalim** [comment]
I am not sure so I will wait for others to opine. Traditionally doctests do not check the notation used because that's not its job. Its job is to run the code inside the notation and it can do that without those limitations today. I also think it would be complicated to do complex checks. Sure we can check it does not start with `...>` but it would be hard to check if `...>` is being used correctly because it implies an incomplete sequence in the previous line, which you can only do after tokenizing (and such would add to the time that takes to compile and run doctests).

**Comment by eksperimental** [comment]
that's exactly because I can see it won't be that easy to implement, but if any of those checks are considered necessary, not a problem implementing them.
so lets wait for other people's opinions

**Comment by whatyouhide** [comment]
Personally, I love strictness and I am not a fan of `iex(1)>` support either (unless I'm missing some reason why we support it other than convenience). It it were up to me, I'd just force users to use `iex>` and `...>`, which would reduce the number of variable elements the Elixir code written in the wild, which is something I enjoy 😄

**Comment by eksperimental** [comment]
@whatyouhide but that would make doctests backward incompatible. Still, that would reduce the number of cases where could would be considered invalid.

**Comment by eksperimental** [comment]
in the meantime I have updated the PR, please have a look at it.
It implements the error message and I had to refactor the extract_tests function a bit.

**Comment by whatyouhide** [comment]
@eksperimental yes, it would make the doctests backwards incompatible, and that's why we can't really do it for now. And this is exactly the point why I am 👎 on supporting the host: then we will need to maintain it and we'll not be able to remove it without backwards-incompatible changes. I still don't see any valid use case for the host.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 1]
```diff
@@ -451,7 +454,8 @@ defmodule ExUnit.DocTest do
     end
   end
 
-  defp adjust_indent(kind, [line | rest], line_no, adjusted_lines, indent, module) when kind in [:prompt, :after_prompt] do
+  defp adjust_indent(kind, [line | rest], line_no, adjusted_lines, indent, module)
+      when kind in [:prompt, :after_prompt] do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Should be aligned to identifier.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 2]
```diff
@@ -428,7 +431,7 @@ defmodule ExUnit.DocTest do
   defp extract_tests(line_no, doc, module) do
     all_lines = String.split(doc, ~r/\n/, trim: false)
     lines = adjust_indent(all_lines, line_no + 1, module)
-    extract_tests(lines, "", "", [], true)
+    do_extract_tests(lines, "", "", [], true, module)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Do we need the `do_` prefix here? If we don't, we shouldn't add it.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 3]
```diff
@@ -509,84 +514,96 @@ defmodule ExUnit.DocTest do
...
+  defp do_extract_tests([{"...>" <> string, _} | lines], expr_acc, expected_acc, acc, newtest, module)
+      when expr_acc != "" do
+    do_extract_tests(lines, expr_acc <> "\n" <> string, expected_acc, acc, newtest, module)
   end
 
   # Expression numbers are simply skipped.
-  defp extract_tests([{<<"iex(", _>> <> string, line} | lines], expr_acc, expected_acc, acc, newtest) do
-    extract_tests([{"iex" <> skip_iex_number(string), line} | lines], expr_acc, expected_acc, acc, newtest)
+  defp do_extract_tests([{<<"iex(", _>> <> string = line, line_no} | lines],
+      expr_acc, expected_acc, acc, newtest, module) do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Should be indented below the first argument.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 4]
```diff
@@ -509,84 +514,96 @@ defmodule ExUnit.DocTest do
...
-  defp skip_iex_number(")>" <> string) do
+  defp skip_iex_number(")>" <> string, _module, _line_no, _line) do
     ">" <> string
   end
 
-  defp skip_iex_number(<<_>> <> string) do
-    skip_iex_number(string)
+  defp skip_iex_number("", module, line_no, line) do
+    raise Error, line: line_no, module: module,
+                 message: ~s{unknown IEx prompt: "#{line}".\n} <>
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Instead of `"#{line}"` it should be `#{inspect line}`.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 5]
```diff
@@ -210,9 +210,19 @@ end |> write_beam
 
 defmodule ExUnit.DocTestTest.Numbered do
   @doc """
-  iex(1)> 1 +
-  ...(1)> 2
-  3
+      iex(1)> 1 +
```
Comment (josevalim) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: Why this change?

**Comment by whatyouhide** [comment]
> In Elixir, the rule is to indent under the same type.

I am quite scared by this rule. What happens with `defmacrop`? Do we indent *10* spaces? 😕 I thought the rule was just to indent `when` with 4 spaces to the right of its `def*`.

Also small side note, technically `foo` and `bar` in `def foo when bar` are LHS and RHS of the `when` operator but isn't that an implementation detail? I don't really see `when` as an operator but maybe that's just me :)

**Comment by josevalim** [comment]
I don't think `when` being an operator is an implementation detail. It is part of the language syntax.

Also, what is the issue with indenting with 10 spaces? You are not going to run out of vertical space.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 5 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: oops

**Comment by eksperimental** [comment]
same here, as @whatyouhide I thought the rule was 2 indentations levels, therefore 4 spaces.
i don't have a problem with going X spaces to the right, the only nuisance is that code may end up indented at an odd number of characters and then you cannot read that by hitting <kbd>Tab</kbd>.

**Comment by whatyouhide** [comment]
I don't have a problem with indenting *one line* 10 spaces, I just am not a fan of the blank rectangles of whitespace that tend to form when a lot of stuff is indented by a lot. Something along the lines of [this comment of yours](https://github.com/levionessa/elixir_style_guide/pull/33#issuecomment-76914655) @josevalim :)

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 2 above)
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: We only do that if we need to implement a private function with exact the same arity. If the arity is not taken, then we go without the `do_`. And @lexmag and I have been discussing to find ways to avoid the `do_` altogether and with with more describe names, such as `extract_tests_with_acc`.

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 4 above)
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Btw, the reason why we should align `when` with the previous "type" is exactly why this line is indented to N spaces and not 2.

**Comment by josevalim** [comment]
@whatyouhide if you want to break it to a new line and use two spaces, fine. But if you are breaking in the middle of it, then it should be aligned accordingly. See the example above with `raise Error` and we use the exact same convention of aligning to the previous type.

**Comment by whatyouhide** [comment]
@josevalim ok, then `when` shall be indented same level as the function/macro being defined 👍

> the only nuisance is that code may end up indented at an odd number of characters and then you cannot reach that by hitting Tab

@eksperimental that's really an editor problem though 😄

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 4 above)
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: just to double check for future cases. we know for sure line is a string, because it comes from String.split/ we still use `inspect`, right?

**Comment by whatyouhide** [comment]
(@josevalim https://github.com/lexmag/elixir-style-guide/pull/9 😄)

**Comment by eksperimental** [comment]
please have a look again at everything,
thank you for all your input

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 4 above)
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: We use inspect so we wrap it in double quotes and we convert new lines into \n, tabs into \t and so on.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 6]
```diff
@@ -509,84 +514,96 @@ defmodule ExUnit.DocTest do
...
-  defp skip_iex_number(")>" <> string) do
+  defp skip_iex_number(")>" <> string, _module, _line_no, _line) do
     ">" <> string
   end
 
-  defp skip_iex_number(<<_>> <> string) do
-    skip_iex_number(string)
+  defp skip_iex_number("", module, line_no, line) do
+    raise Error, line: line_no, module: module,
+                 message: ~s{unknown IEx prompt: #{inspect line}.\n} <>
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: I think we can define the message in a previous line and then call error:

```elixir
message = "unknown IEx prompt: #{inspect line}. Accepted formats are: iex>, iex(1)>, ...>, ...(1)>"
raise Error, line: line_no, module: module, message: message
```

I think it will be clear than nesting as today.

Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: done

**Comment by eksperimental** [comment]
rebased.

**Comment by josevalim** [comment]
Approved.

**Comment by whatyouhide** [comment]
I am a bit conflicted for all the unrelated and maybe needless changes in this PR (such as splitting long lines or replacing `~r` with `~R` even when unnecessary) 😕

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 7]
```diff
@@ -384,7 +384,7 @@ defmodule ExUnit.DocTestTest do
     output = capture_io(fn -> ExUnit.run end)
 
     # Test order is not guaranteed, we can't match this as a string for each failing doctest
-    assert output =~ ~r/\d+\) test moduledoc at ExUnit\.DocTestTest\.Invalid \(\d+\) \(ExUnit\.DocTestTest\.ActuallyCompiled\)/
+    assert output =~ ~R/\d+\) test moduledoc at ExUnit\.DocTestTest\.Invalid \(\d+\) \(ExUnit\.DocTestTest\.ActuallyCompiled\)/
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: I'm wondering why we replacing one with another?

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 8]
```diff
@@ -373,7 +383,7 @@ defmodule ExUnit.DocTestTest do
   test "doctest failures" do
     # When adding or removing lines above this line, the tests below will
     # fail because we are explicitly asserting some doctest lines from
-    # ActuallyCompiled in the format of test/ex_unit/doc_test_test.exs:<LINE>.
+    # ActuallyCompiled in the format of: test/ex_unit/doc_test_test.exs:<LINE>
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: We usually have period in comments.

Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: Probably `ActuallyCompiled in the format of "test/ex_unit/doc_test_test.exs:<LINE>"` would be even better than adding colon.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 9]
```diff
@@ -507,6 +517,22 @@ defmodule ExUnit.DocTestTest do
     assert capture_io(fn -> ExUnit.run end) =~ "1 test, 0 failures"
   end
 
+  test "IEx prompt contains host" do
+    msg =
+      ~S{unknown IEx prompt: "iex(foo@bar)1> 1 +".} <> "\n" <>
+      ~S{Accepted formats are: iex>, iex(1)>, ...>, ...(1)>}
+    regex = ~r[test/ex_unit/doc_test_test\.exs:\d+: #{Regex.escape(msg)}]
+
+    assert_raise ExUnit.DocTest.Error,
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: I think current formatting is harder to comprehend than:
```elixir
assert_raise ExUnit.DocTest.Error, regex, fn ->
  defmodule HostUsage do
    use ExUnit.Case
    doctest ExUnit.DocTestTest.Host
  end
end
```

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 10]
```diff
@@ -507,6 +517,22 @@ defmodule ExUnit.DocTestTest do
     assert capture_io(fn -> ExUnit.run end) =~ "1 test, 0 failures"
   end
 
+  test "IEx prompt contains host" do
+    msg =
+      ~S{unknown IEx prompt: "iex(foo@bar)1> 1 +".} <> "\n" <>
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: I believe we use `[` more frequently than `{` in similar cases.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 11]
```diff
@@ -507,6 +517,22 @@ defmodule ExUnit.DocTestTest do
     assert capture_io(fn -> ExUnit.run end) =~ "1 test, 0 failures"
   end
 
+  test "IEx prompt contains host" do
+    msg =
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: It doesn't hurt to use `message` here (we already do this in the place where we raise the message). :bowtie:

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 10 above)
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: I think in general we favour lower case sigils over upper ones.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 12]
```diff
@@ -511,91 +519,107 @@ defmodule ExUnit.DocTest do
...
   end
 
-  defp extract_tests([], "", "", acc, _) do
+  defp extract_tests([], "", "", acc, _, _) do
     Enum.reverse(acc)
   end
 
   # End of input and we've still got a test pending.
-  defp extract_tests([], expr_acc, expected_acc, [test | t], _) do
+  defp extract_tests([], expr_acc, expected_acc, [test | acc_rest], _, _) do
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: `acc_rest` seems confusing, we use either `acc` or `rest` in other places.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 13]
```diff
@@ -511,91 +519,107 @@ defmodule ExUnit.DocTest do
...
     ">" <> string
   end
 
-  defp skip_iex_number(<<_>> <> string) do
-    skip_iex_number(string)
+  defp skip_iex_number("", module, line_no, line) do
+    message =
+      "unknown IEx prompt: #{inspect line}.\nAccepted formats are: iex>, iex(1)>, ...>, ...(1)>}"
+    raise Error, line: line_no, module: module,
+                 message: message
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: It could go on previous line without loosing readability.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 7 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: there is no interpolation, so why do we want to use ~r. we are making an unnecessary call here everytime we do that.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 8 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: we can do that, I just thought is less noise when you use colons at the end of the line

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 9 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: thank you @lexmag

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 12 above)
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: thank you.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 10 above)
Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: curly braces replaced.
regarding your second comment, I have replied above why i favour uppercase sigils.

**Comment by eksperimental** [comment]
@lexmag almost all your suggestions have been addressed in the last commit

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 7 above)
Comment (josevalim) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: @eksperimental using `~r` or `~R` has no difference at compile or runtime.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 14]
```diff
@@ -445,13 +448,15 @@ defmodule ExUnit.DocTest do
   defp adjust_indent(:text, [line | rest], line_no, adjusted_lines, indent, module) do
     case String.starts_with?(String.trim_leading(line), @iex_prompt) do
       true  ->
-        adjust_indent(:prompt, [line | rest], line_no, adjusted_lines, get_indent(line, indent), module)
+        adjust_indent(:prompt, [line | rest], line_no, adjusted_lines, get_indent(line, indent),
+                      module)
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Seems unnecessary change to me. :bowtie:

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 15]
```diff
@@ -485,11 +491,13 @@ defmodule ExUnit.DocTest do
     stripped_line = strip_indent(line, indent)
     cond do
       stripped_line == "" ->
-        adjust_indent(:text, rest, line_no + 1, [{stripped_line, line_no} | adjusted_lines], 0, module)
+        adjust_indent(:text, rest, line_no + 1, [{stripped_line, line_no} | adjusted_lines],
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: I assume previous length was acceptable here.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 16]
```diff
@@ -485,11 +491,13 @@ defmodule ExUnit.DocTest do
...
+                      0, module)
       String.starts_with?(String.trim_leading(line), @iex_prompt) ->
         adjust_indent(:prompt, [line | rest], line_no, adjusted_lines, indent, module)
       true ->
-        adjust_indent(:code, rest, line_no + 1, [{stripped_line, line_no} | adjusted_lines], indent, module)
+        adjust_indent(:code, rest, line_no + 1, [{stripped_line, line_no} | adjusted_lines],
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: I assume previous length was acceptable here.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 17]
```diff
@@ -466,7 +471,8 @@ defmodule ExUnit.DocTest do
           else: "#{indent} spaces"
 
         raise Error, line: line_no, module: module,
-                     message: "indentation level mismatch: #{inspect line}, should have been #{n_spaces}"
+                     message: "indentation level mismatch: #{inspect line}, " <>
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: With addition of `<>` operation this whole construction starts to be overloaded with actions.
I think moving this to the:
```elixir
message =
  "indentation level mismatch: #{inspect line}, " <>
  "should have been #{n_spaces}"
```
will improve situation.

Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Otherwise addition of `[]` around options for `raise` call.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 18]
```diff
@@ -511,91 +519,105 @@ defmodule ExUnit.DocTest do
...
     test = add_expr(test, expr_acc, expected_acc)
-    Enum.reverse([test | t])
+    Enum.reverse([test | rest])
   end
 
   # We've encountered the next test on an adjacent line. Put them into one group.
-  defp extract_tests([{"iex>" <> _, _} | _] = list, expr_acc, expected_acc,
-                     [test | t], newtest) when expr_acc != "" and expected_acc != "" do
+  defp extract_tests([{"iex>" <> _, _} | _] = list, expr_acc, expected_acc, [test | rest],
+                     new_test, module) when expr_acc != "" and expected_acc != "" do
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: I think we can keep `new_test, module)` on previous line.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 19]
```diff
@@ -511,91 +519,105 @@ defmodule ExUnit.DocTest do
...
-    extract_tests(lines, string, expected_acc, acc, false)
+  defp extract_tests([{"iex>" <> string, _} | lines], "", expected_acc, acc, false, module) do
+    extract_tests(lines, string, expected_acc, acc, false, module)
   end
 
   # Still gathering expr_acc. Synonym for the next clause.
-  defp extract_tests([{"iex>" <> string, _} | lines], expr_acc, expected_acc, acc, newtest) do
-    extract_tests(lines, expr_acc <> "\n" <> string, expected_acc, acc, newtest)
+  defp extract_tests([{"iex>" <> string, _} | lines], expr_acc, expected_acc, acc,
+                     new_test, module) do
```
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: I think it is acceptable to keep arguments on the previous line.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 20]
```diff
@@ -280,6 +280,16 @@ defmodule ExUnit.DocTestTest.Numbered do
   def test_fun(), do: :ok
 end |> write_beam()
 
+
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: Unnecessary line. ✂️

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 7 above)
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: @eksperimental as @josevalim said, there is no actual difference, thus I think better to stick with `~r` to lower changes, and we favour lower case sigils over upper ones in general.

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 14 above)
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: trying to fit everything under 100 chars

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 7 above)
Comment (whatyouhide) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: Agreed, let's go with `~r` for less changes and because I believe `~R` raises more questions (why is this uppercase?) than what we get out of it (basically nothing). :)

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 14 above)
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: We don't have this rule in Elixir, and in any case splitting long lines into smaller lines should be done purposefully (I don't believe having 5 args on one line and one on the next is good visually) and not in this PR since it is completely unrelated. :)

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 18 above)
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Agreed with @lexmag.

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 14 above)
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: i will revert these changes i guess

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 17 above)
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: I will squeeze everything in one line of 105 chars

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 14 above)
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: 100 chars seems too strict limit, it might sometimes make things worse I suppose.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs (hunk 7 above)
Comment (whatyouhide) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: These are tests. We're not interested in such small performance tweaks in tests. And also, this happens at compile time anyways (now for test this may be irrelevant, but it seems inefficient to differentiate).

Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: Yes, it does `Macro.unescape_string/2` call at compile time. We all agree to accept that fact and let it be. :bowtie:

**Comment by eksperimental** [comment]
fixed

**Comment by whatyouhide** [comment]
I will not approve these changes myself because I stand against unrelated changes, and there are a gazillion in this PR. They make the final commit feel confusing for someone going through the history as it's hard to find the changes that actually contributed to bring in the feature. If others in the team feel ok with this, feel free to go ahead of course 😃

**Comment by josevalim** [comment]
The only "unrelated" change in the PR at this point is adding the `\.` escape in some of the tests which is a welcome "normalization" and improvement to the tests. Therefore this looks good to go for me.

**Comment by whatyouhide** [comment]
@josevalim sorry, I didn't point out the changes :(, I meant all the long lines split into two lines!

**Comment by josevalim** [comment]
@whatyouhide a new argument was added to those function clauses. We would very likely have done the same if I am adding a new argument and realize the line has now become too long.

**Comment by whatyouhide** [comment]
@josevalim https://github.com/elixir-lang/elixir/pull/5386/files#diff-523da743d62635a44900f16f8cd7f504L286, https://github.com/elixir-lang/elixir/pull/5386/files#diff-523da743d62635a44900f16f8cd7f504L348, https://github.com/elixir-lang/elixir/pull/5386/files#diff-523da743d62635a44900f16f8cd7f504L365. I realize I am terribly guilty of exaggerating with "gazillion" 😄, sorry @eksperimental, but those three changes are just splitting the line for the sake of it. As I said if we're ok with this I'm not opposed to :)

**Comment by josevalim** [comment]
@whatyouhide please drop a comment on those ones so we can address it.  It is indeed important we keep PRs focused.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 21]
```diff
@@ -283,7 +283,8 @@ defmodule ExUnit.DocTest do
         error ->
           original_stack = System.stacktrace
           reraise ExUnit.AssertionError,
-            [message: "Doctest failed: got #{inspect(error.__struct__)} with message " <> inspect(Exception.message(error)),
+            [message: "Doctest failed: got #{inspect(error.__struct__)} with message " <>
+                      inspect(Exception.message(error)),
```
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Unnecessary change IMO, we don't have a line length standard in Elixir and while I agree these lines are quite long, these changes tend to make the PR less focused.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 22]
```diff
@@ -345,7 +346,8 @@ defmodule ExUnit.DocTest do
           message =
             cond do
               actual_exception != unquote(exception) ->
-                "Doctest failed: expected exception #{inspect(unquote(exception))} but got #{inspect(actual_exception)} with message #{inspect(actual_message)}"
+                "Doctest failed: expected exception #{inspect(unquote(exception))} " <>
+                "but got #{inspect(actual_exception)} with message #{inspect(actual_message)}"
```
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Same as the "Unnecessary change" comment above.

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 23]
```diff
@@ -362,7 +364,8 @@ defmodule ExUnit.DocTest do
       else
         _ ->
           reraise ExUnit.AssertionError,
-            [message: "Doctest failed: expected exception #{inspect(unquote(exception))} but nothing was raised",
+            [message: "Doctest failed: expected exception #{inspect(unquote(exception))} " <>
+                      "but nothing was raised",
```
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Same as the "Unnecessary change" comment above.

File: lib/ex_unit/lib/ex_unit/doc_test.ex (hunk 21 above)
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Unnecessary change IMO, we don't have a line length standard in Elixir and while I agree these lines are quite long, these changes tend to make the PR less focused.

File: lib/ex_unit/test/ex_unit/doc_test_test.exs [hunk 24]
```diff
@@ -373,7 +382,7 @@ defmodule ExUnit.DocTestTest do
   test "doctest failures" do
     # When adding or removing lines above this line, the tests below will
     # fail because we are explicitly asserting some doctest lines from
-    # ActuallyCompiled in the format of test/ex_unit/doc_test_test.exs:<LINE>.
+    # ActuallyCompiled in the format of: "test/ex_unit/doc_test_test.exs:<LINE>"
```
Comment (lexmag) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: Please don't remove period in the comment. :bowtie:

Comment (eksperimental) [inline lib/ex_unit/test/ex_unit/doc_test_test.exs]: fixed (semicolon removed as well)

**Comment by eksperimental** [comment]
no worries @whatyouhide. 
sometimes it's a bit of a greyish line on what's a "normalization" and what is not.
all "unrelated" changes have been reverted
I will certainly try to make more focused PRs and make them smaller as well so it helps reviewing them.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart: :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by eksperimental** [comment]
thank you
